keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,535 @@
+
+Pod Topology Spread has the semantics of "global minimum", which means the minimum number of pods that match the label selector in a topology domain.
+
+When the number of domains with matching topology keys is less than `minDomains`,
+Pod Topology Spread treats "global minimum" as 0; other wise, "gloval minimum"
+is equal to the minimum number of matching pods on a domain.
+
+As a result,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same or more number of matching Pods as `maxSkew`.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that have the less number of matching Pods as `maxSkew`.
+
+`minDomains` is an optional parameter and the default value is 0.
+
+### API
+
+New optional parameter called `MinDomains` is introduced to `PodSpec.TopologySpreadConstraint[*]`.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // When the number of domains with matching topology keys is less than `minDomains`,
+  // Pod Topology Spread treats "global minimum" as 0.
+  // As a result,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`,
+  //   scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same number of matching Pods as `maxSkew`.
+  // - when `whenUnsatisfiable` equals `ScheduleAnyway`,
+  //   scheduler prefers Nodes on the domains that don't have the same number of matching Pods as `maxSkew`.
+  // Default value is 0.
+  // +optional
+  MinDomains *int32
+}
+```
+
+### Implementation details
+
+**Filtering phase**
+
+In Filter of Pod Topology Spread, filtering criteria is
+
+```
+('existing matching num' + 'if self-match (1 or 0)' - 'global min matching num') <= 'maxSkew'
+```
+
+- `existing matching num` denotes the number of current existing matching Pods on the domain.
+- `if self-match` denotes if the labels of Pod is match with selector of the constraint.
+- `global min matching num` denotes the minumun number of matching Pods.
+
+For `whenUnsatisfiable: DoNotSchedule`, Pod Topology Spread treats `global min matching num` as 0
+when the number of domains with matching topology keys is less than `minDomains`.
+
+In `preFilterState`, Pod Topology Spread has `Constraints` field which type is slice of internal `topologySpreadConstraint` struct.
+By introducing `MinDomains` field in `topologySpreadConstraint`,
+Pod Topology Spread can use `MinDomains` to determine the value of `global min matching num` when we calculate filtering criteria.
+
+This change will not increase the complexity of filtering logic and the performance degradation is not expected.
+
+**Scoring phase**
+
+When the number of domains with matching topology keys is less than `minDomains` and `whenUnsatisfiable` equals to `ScheduleAnyway`,
+Pod Topology Spread gives low scores to Nodes on the domains which have the same or more number of matching Pods as `maxSkew`.
```

Comment:
low or zero score? Clarify.

Also, it's unclear if you are talking about the current state of the scoring phase or the desired state after you implement the feature.

Reply:
> low or zero score? Clarify.

Okay. My way gives low score. It's not always zero.

> it's unclear if you are talking about the current state of the scoring phase or the desired state after you implement the feature.

Okay. I'll try to change my English wording 🙇